btx/diagnostics/run.py

```diff
@@ -92,28 +103,60 @@ def compute_stats(self, img):
         self.stats['max'][self.n_proc] = np.max(img)
         self.stats['min'][self.n_proc] = np.min(img)
         
-    def finalize_stats(self, n_empty=0):
+    def finalize_stats(self, n_empty=0, n_empty_raw=0):
         """
         Gather stats from various ranks into single arrays in self.stats_final.
 
         Parameters
         ----------
         n_empty : int
             number of empty images in this rank
+        n_empty_raw : int
+            number of empty raw events in this rank
         """
         self.stats_final = dict()
         for key in self.stats.keys():
             if n_empty != 0:
-                self.stats_final[key] = self.stats[key][:-n_empty]
+                self.stats[key] = self.stats[key][:-n_empty]
             self.stats_final[key] = self.comm.gather(self.stats[key], root=0)
 
         self.stats_final['fiducials'] = self.comm.gather(np.array(self.psi.fiducials), root=0)
         if self.rank == 0:
             for key in self.stats_final.keys():
-                #self.stats_final[key] = np.array(self.stats_final[key]).reshape(-1)
                 self.stats_final[key] = np.hstack(self.stats_final[key])
+                
+        if self.gain_traj is not None:
+            if n_empty_raw != 0:
+                self.gain_traj = self.gain_traj[:-n_empty_raw]
+            self.stats_final['gain_mode_counts'] = self.comm.gather(self.gain_traj, root=0)
+            if self.rank == 0:
+                self.stats_final['gain_mode_counts'] = np.hstack(self.stats_final['gain_mode_counts'])
+                
+    def get_gain_statistics(self, raw, gain_mode='AML-L'):
+        """
+        Retrieve statistics for a particular gain mode, specifically a map 
+        of the number of times a pixel has been in a certain gain mode and 
+        the number of pixels in that gain mode per event.
 
-    def compute_run_stats(self, max_events=-1, mask=None, powder_only=False, threshold=None):
+        Parameters
+        ----------
+        raw : ndarray, shape (det_shape)
+            uncalibrated image
+        gain_mode : str
+            gain mode to retrieve statistics for, e.g. 'AML-L'
+        """
+        if self.gain_traj is None:
+            self.gain_mode = gain_mode
+            self.modes = {'FH':0, 'FM':1, 'FL':2, 'AHL-H':3, 'AML-M':4, 'AHL-L':5, 'AML-L':6}
+            self.gain_traj = np.zeros(self.psi.max_events - self.psi.counter).astype(int)
+            self.gain_map = np.zeros(self.psi.det.shape())
+
+        stats = info_pixel_gain_mode_statistics_for_raw(self.psi.det, raw)
+        self.gain_traj[self.n_proc] = int(stats.split(":")[1].split(",")[self.modes[gain_mode]])
+        evt_map = map_pixel_gain_mode_for_raw(self.psi.det, raw)
+        self.gain_map[evt_map==[self.modes[gain_mode]]] += 1
```

Comment:
Since we might have to deal with other detectors like the Jungfrau or different flavors of epix10k in the future, how should we prepare? Should this function be able to deal with any detector or do we want one function per detector, in which case we might want to rename it `get_epix10k2M_gain_statistics`?
I prefer the former approach but maybe the latter is more practical?

Reply:
I agree and think the former approach (with one function for all detectors) is workable.